From 64cc84c209b8c2eaf0fa8ba8cabe842e538f3507 Mon Sep 17 00:00:00 2001
From: Chris Wilson <chris@chris-wilson.co.uk>
Date: Tue, 14 Sep 2010 10:22:23 +0100
Patch-mainline: 2.6.37
References: fate#310916
Git-commit: 0bc23aad3b67ca0cd7480dec0b7652d9b8686432
Subject: [PATCH 1407/2588] drm/i915: Fix regression in ba3d8d749b

I pulled the wrong version of the patch from Daniel Vetter which was
missing the read barriers -- and the one that was causing all the trouble
was from i915_gem_object_put_fence_reg(), leading to GPU hangs on gen3.

Signed-off-by: Chris Wilson <chris@chris-wilson.co.uk>
(cherry picked from commit 0bc23aad3b67ca0cd7480dec0b7652d9b8686432)

Signed-off-by: Takashi Iwai <tiwai@suse.de>
---
 drivers/gpu/drm/i915/i915_gem.c |    8 ++++++--
 1 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_gem.c b/drivers/gpu/drm/i915/i915_gem.c
index 74c6475..7b23d7e 100644
--- a/drivers/gpu/drm/i915/i915_gem.c
+++ b/drivers/gpu/drm/i915/i915_gem.c
@@ -2442,12 +2442,16 @@ i915_gem_object_put_fence_reg(struct drm_gem_object *obj)
 		int ret;
 
 		ret = i915_gem_object_flush_gpu_write_domain(obj, false);
-		if (ret != 0)
+		if (ret)
+			return ret;
+
+		ret = i915_gem_object_wait_rendering(obj);
+		if (ret)
 			return ret;
 	}
 
 	i915_gem_object_flush_gtt_write_domain(obj);
-	i915_gem_clear_fence_reg (obj);
+	i915_gem_clear_fence_reg(obj);
 
 	return 0;
 }
-- 
1.7.6

